datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyMulticastStrategy.java

```diff
@@ -128,35 +162,55 @@ public StickyMulticastStrategy(Optional<Integer> maxTasks, Optional<Integer> imb
         allAliveTasks.addAll(foundDatastreamTasks);
 
         if (!foundDatastreamTasks.isEmpty()) {
-          if (foundDatastreamTasks.size() > numTasks) {
+          if (behaveAsNormalAssignment && (foundDatastreamTasks.size() > numTasks)) {
             LOG.info("Skipping {} tasks from the previous assignment of instance {}.",
                 foundDatastreamTasks.size() - numTasks, instance);
             foundDatastreamTasks = foundDatastreamTasks.stream().limit(numTasks).collect(Collectors.toList());
           }
           newAssignment.get(instance).addAll(foundDatastreamTasks);
           currentAssignmentCopy.get(instance).removeAll(foundDatastreamTasks);
-          numTasks -= foundDatastreamTasks.size();
+          if (behaveAsNormalAssignment) {
+            numTasks -= foundDatastreamTasks.size();
+          }
+          numTasksFound += foundDatastreamTasks.size();
```

Comment:
yup, totally agree. as discussed, persistence of the numTasks to ZK will fix this. This will be done as part of a future change. I'll also keep the in-memory map to reduce our calls to ZK.